app/frontend/src/components/MapSearch.tsx

```diff
@@ -0,0 +1,131 @@
+    top: 10,
+    width: "70%",
+    padding: theme.spacing(1),
+    borderRadius: theme.shape.borderRadius,
+    zIndex: 1,
+    background: theme.palette.background.default,
+    opacity: 0.9,
+    "& *": {
+      opacity: 1,
+    },
+    "& .MuiAutocomplete-input": {
+      fontSize: "0.75rem",
+    },
+    "& .MuiInputLabel-root": {
+      fontSize: "0.75rem",
+    },
+    "& .MuiFormHelperText-root": {
+      fontSize: "0.65rem",
+    },
+  },
+}));
+
+interface SearchOption {
+  name: string;
+  location: LngLat;
+}
+
+interface MapSearchProps {
+  setCity: (value: string) => void;
+  setError: (error: string) => void;
+  setMarker: (lngLat: LngLat) => void;
+}
+
+export default function MapSearch({
+  setCity,
+  setError,
+  setMarker,
+}: MapSearchProps) {
+  const classes = useSearchStyles();
+
+  const [searchOptionsLoading, setSearchOptionsLoading] = useState(false);
+  const [searchOptions, setSearchOptions] = useState<SearchOption[]>([]);
+  const [open, setOpen] = useState(false);
+
+  const loadSearchOptions = async (value: string) => {
+    setError("");
+    setOpen(true);
+    setSearchOptions([]);
+    if (!value) {
+      return;
+    }
+    setSearchOptionsLoading(true);
+    const url =
+      nominatimURL! + "search?format=jsonv2&q=" + encodeURIComponent(value);
+    const options = {
+      method: "GET",
+      headers: {
+        Accept: "application/json",
+        "Content-Type": "application/json;charset=UTF-8",
+      },
+    };
+    try {
+      const res = await fetch(url, options);
+      const data = (await res.json()) as Array<any>;
+      console.log(data);
+      setSearchOptions(
+        data.map((obj) => ({
+          name: obj["display_name"],
+          location: new LngLat(Number(obj["lon"]), Number(obj["lat"])),
+        }))
+      );
+    } catch (e) {
+      setError(e.message);
+      setOpen(false);
+    }
+    setSearchOptionsLoading(false);
+  };
+
+  const searchSubmit = (
+    value: string | SearchOption | null,
+    reason: AutocompleteChangeReason
+  ) => {
+    console.log(reason);
+    if (reason === "blur") {
+      setOpen(false);
+      return;
+    }
+    if (!(value as SearchOption)?.name) {
+      setCity(value as string);
+      //create-option is when enter is pressed on user-entered string
+      if (reason === "create-option") {
+        loadSearchOptions(value as string);
+      }
+    } else {
+      const searchOption = value as SearchOption;
+      setMarker(searchOption.location);
+      setOpen(false);
+    }
+  };
+
+  return (
+    <Autocomplete
+      label="My location"
+      //hint="Press enter to search the map, then customize the text."
+      size="small"
+      options={searchOptions}
+      getOptionLabel={(o) => (!!o.name ? o.name : String(o))}
+      loading={searchOptionsLoading}
+      open={open}
+      onBlur={() => setOpen(false)}
+      onChange={(_, value, reason) => searchSubmit(value, reason)}
+      onInputChange={(_, value) => setCity(value)}
```

Comment:
Had a play around on Storybook and I suspect this is causing state update on every keystroke. Is that really needed considering the search result only shows when the I press enter?

Reply:
I do not think the issue is because of setCity, as that leads to console.log.

Reply:
I just tried and I do not see a rerender while typing.

Reply:
It looks like `setCity` is setting some parent state since it comes from a prop. There's no example anywhere else in the app apart from this storybook entry so I'm not sure where it comes from. But the main point is, I'm not sure we need to capture every keystroke update.

Reply:
You will pass in setCity to update form state (I guess from a Controller). Is there a better way? The problem is, pressing enter performs the search, so that can't really be used as the event for updating the text in the form state.

Reply:
Oh I see, in that case, I think it might be fine then, as I believe react-hooks-form already optimise for those things :+1: